### ISSUE TITLE
CE: Allowing let to Implicitly Use Bind

### ISSUE THREAD

**Issue by SchlenkR** [fsharp/fslang-suggestions#1392]
# Allowing let to Implicitly Use Bind

In the context of Computation Expressions (CEs), it would be beneficial to allow let bindings to implicitly invoke the `Bind` method when available, effectively replacing the need for `let!`.

This change aims to simplify the syntax and improve the developer experience by prioritizing the abstracted result over the underlying type.

## Explanation

Currently, within a computation expression, you use `let!` to bind the result of a computation, leveraging the `Bind` method of the builder. However, developers are often more interested in the result of the computation rather than the specifics of the builder type (I say). By giving precedence to a potential `Bind` overload, `let`-bindings could automatically perform the bind operation when appropriate.

**Example Before Enhancement:**

```fsharp
scene {
    let! agif = Image.load "animated.gif"
    let frameCount = agif.frames.length
    let duration = 5.0 / frameCount
    let! opacity = Animate.linear(1.0, 0.0, duration)
    Image.show(agif.opacity(opacity))
}
```

**Example After Enhancement:**

```fsharp
scene {
    let agif = Image.load "animated.gif"
    let frameCount = agif.frames.length
    let duration = 5.0 / frameCount
    let opacity = Animate.linear(1.0, 0.0, duration)
    Image.show(agif.opacity(opacity))
}
```

In this example, if `Image.load` and `Animate.linear`return a type that the `scene` builder can bind, the compiler would implicitly use that `Bind` overload for let.

To enable this behavior, the `Bind` method could be defined with an attribute that gives it precedence:

```fsharp
type SceneBuilder() =
    member _.Bind([<LetPrecedenceOverLetBang>] m: 'A, f) = ...
```

The `[<LetPrecedenceOverLetBang>]` attribute signals to the compiler that this Bind method should be considered when resolving let bindings within computation expressions when the type of the let-expr to be bound is inferred to be of type `'A`.

## Benefits

* Simplified Syntax: Developers can write more concise and readable code by using let instead of let! in computation expressions.

* Improved Developer Experience: Reduces the cognitive load of remembering when to use let versus let!, making the language more approachable for newcomers and reducing potential errors.

## Additional Considerations

* Explicit Non-Bind Results: If access to the unbound result is necessary, an alternative syntax like `let!!` (or `let*`, etc.) could be introduced to explicitly bypass the implicit Bind.

* Future Language Design: While this proposal is based on personal experience with F#, it might serve as a useful consideration for future versions of F# or similar languages.

By implementing this enhancement, the language would prioritize the developer’s intention to work with the abstracted results of computations, streamlining code and improving overall usability.

**Comment by SchlenkR** [comment]
...also here:

https://bsky.app/profile/schlenkr.bsky.social/post/3lbut4yui7s24

**Comment by brianrourkeboll** [comment]
I think you'd quickly run into overload resolution ambiguities, among other issues. (Or, at best, perhaps an arms race of overload resolution priorities... Speaking of which, do we have a specific issue for tracking support for [`OverloadResolutionPriorityAttribute`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.overloadresolutionpriorityattribute?view=net-9.0)?)

I also don't think this would be possible to add to existing builders in a backwards-compatible way, unless you somehow made it only take effect if a given usage of a given builder didn't already use any `!` variant or something. Otherwise, examples like this would no longer behave the same:

```fsharp
open System.Threading.Tasks

task {
    let t = backgroundTask { printfn "Started task 1."; do! Task.Delay 1000 done; printfn "Finished task 1." }
    do! task { printfn "Started task 2."; do! Task.Delay 100 done; printfn "Finished task 2." }
    do! t
}

// Started task 1.
// Started task 2.
// Finished task 2.
// Finished task 1.
```

**Comment by SchlenkR** [comment]
Thanks for your answer.

* **Backward Compatibility:** It is a opt-in feature that does not affect existing builders. It is a choice, left to the author of a system, that might be beneficial for certain "kinds" of computations (see below).

* **The Example:** Maybe choosing `async` for showing syntax was not good. Let me clarify: I'm not proposing to change `async` or similar. **I changed the original text / example of this proposal to make this clearer**.

## Is this a reasonable feature?

Async is idiomatic and well-known, and it's clear when dealing with async functions - often due to using a single module, or using identifiers that have "async" or similar in their name. This is not always the case. There's a difference between CEs that focus a single computation type (like `async`), and others - like shown in the example - having a vocabulary that is more wide-spread and not organised in a single module.

Therefore, it is much harder to understand if one is dealing with a value of the computation type or not. And: It is just not necessary to understand! (at least not at that point). In F#, choosing a builder instance is already the explicit statement for abstracting aspects that the builder then handles. The syntax in this proposal reflects the decision for that abstraction.

In other words: The answer to the question "When to bang or not" is of rare interest; it is already answered. One fact that backs this assumption is that once a builder is chosen, "bang" is the almost-always standard case for dealing with values of the underlying computation type(s), but syntactically, it's a special case.

## Implementation

Might be very hard to implement, I wouldn't doubt that.

**Comment by vzarytovskii** [comment]
> Speaking of which, do we have a specific issue for tracking support for OverloadResolutionPriorityAttribute

I made a note for it in the issue asking for overload tiebreakers

**Comment by vzarytovskii** [comment]
It pretty much was decided as part of 1.0 as key language design choice, to make binding explicit. 

As it was mentioned already, it will bring not only confusion, but numerous overload resolution issues, possible backwards binary incompatibilities (due to old compilers not understanding new attribute), and more importantly - user confusion.

I would personally be not in favour of introducing it even as opt-in feature.

**Comment by SchlenkR** [comment]
> and more importantly - user confusion.

Perhaps this statement should be viewed with more nuance. From a global perspective, it’s certainly true that things don’t get easier when the default syntax is one way in some cases and another way in others. However, once being within a CE, the opposite is true: it brings clarity, specifically because the recurring question of whether a “bang” is necessary or not simply no longer arises.

**Comment by vzarytovskii** [comment]
> > and more importantly - user confusion.
> 
> 
> 
> Perhaps this statement should be viewed with more nuance. From a global perspective, it’s certainly true that things don’t get easier when the default syntax is one way in some cases and another way in others. However, once being within a CE, the opposite is true: it brings clarity, specifically because the recurring question of whether a “bang” is necessary or not simply no longer arises.

I look at it from the perspective of "all the examples and books I saw use bang for bonding this, as do use!, match!, while! and do!, so why is let special, but only in certan CEs and in certain situations".

**Comment by SchlenkR** [comment]
I’m not sure I fully understand all aspects, so let me recap what I’ve grasped:

1. Should constructs like `use!`, `match!`, etc., behave the same as `let!`, implying that the “bang” could be omitted in all such cases under this proposal?
2. Does it introduce confusion when let and let! differ in some computation expressions (CEs)?

**On Point 1:**

Under this proposal, `use!`, `match!`, `while!`, and `do!` could function like `let!`, making `let` no longer a special case.

**On Point 2:**

I agree that this could indeed be confusing (“From a global perspective, it’s certainly true that things don’t get easier when the default syntax is one way in some cases and another way in others.”)

This introduces a “cost,” but the potential benefit would be greater clarity via a simplification ("it brings clarity, specifically because the recurring question of whether a “bang” is necessary or not simply no longer arises.").

From my personal perspective, I believe the benefit outweighs the cost ("cost" here not in the sense of implementation cost). Of course, I understand and respect that others might have a completely different view—there are many valid arguments against this proposal.

**A Small Addition:**

The existing computation expressions available today are often highly specialized, unique, and different from each other. Yet, they remain intuitively usable despite their diversity.

**Comment by T-Gro** [comment]
> Speaking of which, do we have a specific issue for tracking support for [`OverloadResolutionPriorityAttribute`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.overloadresolutionpriorityattribute?view=net-9.0)?)
> 

I am not aware of an existing proposal.

Adapting it to F# (incl. type inference and respective F#'s different typechecking codepath based on method group size being 1 or >1) will need specific design. This might be further complicated by semantics of the attribute - it is not a .NET attribute, it is a C# attribute which is built as an update to C# method resolution specs. 

The way this is added into the C# method resolution spec is not compatible with F# from it's first sentence (argument list identified) because of type inference, and also because of different rules for automatic conversions.


This is not to say it's impossible, it's more to say this needs an F#-targeted proposal first and unlike other interop scenarios, "Do what C# does" unfortunately cannot be taken here as an approach.

**Comment by SchlenkR** [comment]
> when the type of the let-expr to be bound is inferred to be of type 'A.

meaning, having this:

`let {pat} = {exp} in {body}`

is:

(Only) when the type of {exp} can be inferred to be a "bindable" type with a matching overload on the builder, then this overload is used for binding.

In the example

```fsharp
let f f1 f2 =
    b {
        let xf1 = f1 15
        let! xf2 = f2 42
        return (xf1+xf2)
    }
```

there is either a type annotation required for `f1`, or `let!` must be used explicitly to bind `xf1` (which still could be possible).

**Comment by Tarmil** [comment]
Wouldn't it make more sense and be simpler wrt overload resolution to make this a new method eg `builder.Let(x, f)`?

**Comment by SchlenkR** [comment]
Thanks Tarmil. This would be a possibility for implementing this. 

My personal focus is targeting the question weather or under which circumstances this idea - simplification of a 80/90+ % case is a valuable thing, and in weighting the pros and cons in discussion or reasoning. 

I close this as probably not an F# thing. 

Thanks for your time.
